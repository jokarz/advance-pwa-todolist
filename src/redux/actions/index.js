//import axios from '../../utility/axiosInstance';
import { getData, setData } from '../../utility/localStorageHelper'
//-----------------------------View Archive---------------------$
export const setArchive = data => {
  return {
    type: 'setArchive',
    data
  }
}


//-----------------------------SUCCESS---------------------
export const resetSuccess = () => {
  return {
    type: 'resetSuccess'
  }
}

export const success = () => {
  return {
    type: 'success'
  }
}

export const successAt = data => {
  return {
    type: 'successAt',
    data
  }
}

//-----------------------LOADING---------------------------
export const loadStart = () => {
  return {
    type: 'loadStart'
  }
}

export const loadStop = () => {
  return {
    type: 'loadStop'
  }
}

//---------------------------------------ALERT----------------------
export const alertStart = () => {
  return {
    type: 'alertStart'
  }
}

export const alertStop = () => {
  return {
    type: 'alertStop'
  }
}

export const alertMsg = data => {
  return {
    type: 'updateAlertMsg',
    data
  }
}

export const alertAt = data => { //topic, todo+id, edit+id
  return {
    type: 'alertAt',
    data
  }
}


//-----------------------LOAD TOPIC AND TODOS--------------------------------
export const loadAll = data => {
  return {
    type: 'loadAll',
    data
  }
}


export const autoGetAll = () => {
  return async dispatch => {
    const data = getData()
    dispatch(loadAll(data))
  }
}

export const getAll = () => {
  return async dispatch => {
    const data = getData()
    dispatch(loadAll(data))
  }
}

export const addTopic = title => {
  return async dispatch => {
    title = title.trim()
    const data = getData()
    let index = data.findIndex(item => item.title === title)
    dispatch(loadStart());
    if (title === '') {
      dispatch(loadStop())
      dispatch(alertAt('topic'))
      dispatch(alertMsg('Please enter a valid topic'))
      dispatch(alertStart())
      setTimeout(() => { dispatch(alertStop()) }, 3000)
    } else if (index > -1) {
      dispatch(loadStop())
      dispatch(alertAt('topic'))
      dispatch(alertMsg(`Topic already exist ${data[index].archive ? 'in archived' : ''}`))
      dispatch(alertStart())
      setTimeout(() => { dispatch(alertStop()) }, 3000)
    } else {
      const id = Math.random().toString(36).substring(2)
      setData([{ id, title, archive: false, todos: { "completed": [], "incomplete": [] } }, ...data])
      dispatch(loadStop())
      dispatch(success())
      dispatch(successAt('topic' + title))
      dispatch(getAll())
    }
  }
}

export const archiveTopic = (topicId, archive) => {
  return async dispatch => {
    const data = getData()
    dispatch(loadStart());
    dispatch(resetSuccess());
    let index = data.findIndex(item => item.id === topicId)
    if (index > -1) {
      data[index].archive = archive
      let tempo = data[index]
      data.splice(index, 1)
      data.unshift(tempo)
      if (!data.some(item => item.archive)) {
        dispatch(setArchive(false))
      }
    }
    setData(data)
    dispatch(loadStop())
    dispatch(successAt('topicarchive' + topicId))
    dispatch(success())
    dispatch(getAll())
    dispatch(resetSuccess());
  }
}

export const renameTopic = (topicId, title) => {
  return async dispatch => {
    const data = getData()
    let index = data.findIndex(item => item.title === title)
    dispatch(loadStart());
    dispatch(resetSuccess());
    if (index > -1) {
      dispatch(alertAt('topictitle' + topicId))
      dispatch(alertMsg(`Topic already exist ${data[index].archive ? 'in archived' : ''}`))
      dispatch(alertStart())
      setTimeout(() => { dispatch(alertStop()) }, 3000)
      dispatch(loadStop())
    } else {
      index = data.findIndex(item => item.id === topicId)
      if (index > -1) {
        data[index].title = title.trim()
      }
      setData(data)
      dispatch(getAll())
      dispatch(loadStop())
      dispatch(success())
      dispatch(successAt('topictitle' + topicId))
    }
    setTimeout(() => {
      dispatch(resetSuccess());
    }, 50)

  }
}

export const duplicateTopic = topicId => {
  return async dispatch => {
    const data = getData()
    dispatch(loadStart());
    dispatch(resetSuccess())
    let dup = null
    for (let i = 0; i < data.length; i++) {
      let item = data[i]
      if (item.id === topicId) {
        dup = JSON.parse(JSON.stringify(item))
        break
      }
    }
    let index = 1
    while (true) {
      let exist = false
      let title = `(${index}) ${dup.title}`

      for (let i = 0; i < data.length; i++) {
        if (data[i].title === title) {
          exist = true
          break
        }
      }
      if (!exist) {
        dup.title = title
        dup.id = Math.random().toString(36).substring(2)
        data.unshift(dup)
        break;
      }
      index++
    }
    setData(data)
    dispatch(loadStop())
    dispatch(success())
    dispatch(successAt('topicduplicate' + topicId))
    dispatch(getAll())
    setTimeout(() => {
      dispatch(resetSuccess());
    }, 50)
  }
}

export const deleteTopic = topicId => {
  return async dispatch => {
    const data = getData()
    dispatch(loadStart());
    let index = data.findIndex(item => item.id === topicId)
    if (index > -1) {
      data.splice(index, 1)
      if (!data.some(item => item.archive)) {
        dispatch(setArchive(false))
      }
      setData(data)
    }
    dispatch(loadStop())
    dispatch(success())
    dispatch(getAll())
  }
}

//this part onwards is done differently from the server sided ver
export const addTodo = (topicId, content) => {
  return async dispatch => {
    const data = getData()
    dispatch(loadStart());
    dispatch(resetSuccess());
    let index = data.findIndex(item => item.id === topicId)
    if (index === -1) {
      dispatch(loadStop())
      return
    }
    const allTodos = [...data[index].todos.completed, ...data[index].todos.incomplete]
    if (allTodos.some(todo => todo.content === content.trim())) {
      dispatch(alertAt('todo' + topicId))
      dispatch(alertMsg('Todo already exist~'))
      dispatch(alertStart())
      setTimeout(() => { dispatch(alertStop()) }, 3000)
      dispatch(loadStop())
    } else {
      const newTodo = { id: Math.random().toString(36).substring(2), content }
      data[index].todos.incomplete.unshift(newTodo)
      setData(data)
      dispatch(loadStop())
      dispatch(successAt('addtodo' + topicId))
      dispatch(success())
      dispatch(getAll())
      setTimeout(() => {
        dispatch(resetSuccess());
      }, 50)
    }
  }
}

export const editTodos = (topicId, contents, action) => { // action = ['completed', 'incomplete', 'delete']
  return async dispatch => {
    const data = getData()
    dispatch(loadStart());
    dispatch(resetSuccess());
    let index = data.findIndex(item => item.id === topicId)
    if (index === -1) {
      dispatch(loadStop())
      return
    }

    if (action === 'incomplete' || action === 'completed') {
      let originType = action === 'incomplete' ? 'completed' : 'incomplete'
      contents.forEach(content => {
        let itemIndex = data[index]['todos'][originType].findIndex(item => item.content === content.trim())
        if (itemIndex === -1) {
          dispatch(loadStop())
          return
        }
        let item = data[index]['todos'][originType][itemIndex]
        data[index]['todos'][action].unshift({ ...item })
        data[index]['todos'][originType].splice(itemIndex, 1)
      });
    } else {
      contents.forEach(content => {
        const newIncomplete = data[index]['todos']['incomplete'].filter(todo => todo.content !== content.trim())
        const newCompleted = data[index]['todos']['completed'].filter(todo => todo.content !== content.trim())
        data[index]['todos']['incomplete'] = newIncomplete
        data[index]['todos']['completed'] = newCompleted
      })
    }
    setData(data)
    dispatch(loadStop())
    dispatch(successAt('edittodo' + topicId))
    dispatch(success())
    dispatch(getAll())
  }
}
