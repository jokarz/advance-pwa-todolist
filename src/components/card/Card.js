import React, {useState, useEffect } from 'react'
import { css } from 'aphrodite';
import { connect } from 'react-redux';

import CardTitle from './CardTitle'
import { editTodos } from '../../redux/actions';
import Checkbox from '../checkbox/Checkbox';
import AddTodoInput from '../input/AddTodoInput';
import styles from '../../utility/styles';

const Card = props => {
  const { className, title, completed, incomplete, entrance, id, success, successAt, editTodos, alert, alertAt, alertMessage, viewArchive, setHideFab } = props

  const [incompleteList, setIncompleteList] = useState({})
  const [incompleteEditShow, setIncompleteEditShow] = useState(false)
  const [completedList, setCompletedList] = useState({})
  const [completedEditShow, setCompletedEditShow] = useState(false)

  const [currentTab, setCurrentTab] = useState('incomplete') //incomplete, completed

  const resetIncompleteList = () => {
    let tempIncompleteList = {}
    incomplete.forEach(item => {
      tempIncompleteList[item.content] = false
    });
    setIncompleteList(tempIncompleteList)
  }

  const resetCompletedList = () => {
    let tempCompletedList = {}
    completed.forEach(item => {
      tempCompletedList[item.content] = false
    });
    setCompletedList(tempCompletedList)
  }

  const resetAllList = () => {
    resetCompletedList()
    resetIncompleteList()
  }


  const handleIncompleteClick = (todoContent, state) => {
    let temp = { ...incompleteList }
    temp[todoContent] = state
    setIncompleteList(temp)
  }

  const handleCompletedClick = (todoContent, state) => {
    let temp = { ...completedList }
    temp[todoContent] = state
    setCompletedList(temp)
  }

  const handleMarkList = (action) => {
    let contents = []
    if (incompleteEditShow) {
      contents = Object.keys(incompleteList).filter(item => incompleteList[item])
    } else if (completedEditShow) {
      contents = Object.keys(completedList).filter(item => completedList[item])
    }
    editTodos(id, contents, action)

  }

  useEffect(() => {
    resetAllList()
    // eslint-disable-next-line
  }, [entrance])

  useEffect(() => {
    if (completed.length === 0 && currentTab === 'completed') {
      setCurrentTab('incomplete')
    }
    if (incompleteEditShow) {
      let temp = { ...incompleteList }
      let changed = incomplete.filter(item => Object.keys(incompleteList).filter(marked => incompleteList[marked]).includes(item.content)).map(item => item.content)
      Object.keys(incompleteList).forEach(marked => {
        if (!changed.includes(marked)) {
          temp[marked] = false
        }
      })
      setIncompleteList(temp)
    }

    if (completedEditShow) {
      let temp = { ...completedList }
      let changed = completed.filter(item => Object.keys(completedList).filter(marked => completedList[marked]).includes(item.content)).map(item => item.content)
      Object.keys(completedList).forEach(marked => {
        if (!changed.includes(marked)) {
          temp[marked] = false
        }
      })
      setCompletedList(temp)
    }
    // eslint-disable-next-line
  }, [incomplete, completed])


  useEffect(() => {
    let show = Object.keys(incompleteList).some(key => incompleteList[key])
    setIncompleteEditShow(show);
  }, [incompleteList])

  useEffect(() => {
    let show = Object.keys(completedList).some(key => completedList[key])
    setCompletedEditShow(show);
  }, [completedList])

  useEffect(() => {
    currentTab === 'incomplete' ? resetCompletedList() : resetIncompleteList()
    // eslint-disable-next-line
  }, [currentTab])

  useEffect(() => {
    if (success && successAt === `edittodo${id}`) {
      resetAllList()
    }
    // eslint-disable-next-line
  }, [success, successAt])

  return (
    <>
      <div className={`d-block d-sm-inline-block ${className} `} style={{ width: '100%' }}>
      {/* opacity: show ? 1 : 0,  */}
        <CardTitle id={id} title={title} />
        <div className={`card shadow-sm border-0 ${css(styles.radius)}`}>
          <div className="card-body pt-0 pb-2">
          </div>
          {completed.length ?
            <div className="card-body pb-0">
              <ul className="nav nav-pills">
                <li className="nav-item mr-2">
                  <span className={`select-none  nav-link ${css(styles.radius, styles.tab, currentTab === 'incomplete' ? styles.tabActive : '')}`}
                    onClick={() => setCurrentTab('incomplete')}>Todo!</span>
                </li>

                <li className="nav-item">
                  <span className={`select-none nav-link ${css(styles.radius, styles.tab, currentTab === 'completed' ? styles.tabActive : '')}`}
                    onClick={() => setCurrentTab('completed')}>Completed!</span>
                </li>
              </ul>
            </div> : viewArchive ?  <div className="py-2"></div> : null
          }
          {currentTab === 'incomplete' ?
            <>
              {!viewArchive ?
                <AddTodoInput className={completed.length ? '': 'pt-4'} id={id} countList={incomplete.length + completed.length} setHideFab={setHideFab}/> : null
              }

              {incomplete.length ?
                <div className={`card-body px-0 ${incompleteEditShow ? '' : 'pb-0'}`}>
                  <div className={`px-4 pt-0 ${css(styles.limitHeight)} ${incompleteEditShow ? '' : 'mb-4'}`}>
                    {incomplete.map(item => (<Checkbox key={item.id} id={item.id} topicId={id} text={item.content} clickOption={handleIncompleteClick} viewArchive={viewArchive} />))}
                  </div>
                </div>
                :
                <div className={`card-body d-flex flex-column justify-content-center align-items-center`}>
                  {completed.length ?
                    <>
                      <h1 className="fas fa-glass-cheers text-muted text-center">
                      </h1>
                      <h5 className="font-weight-light m-0 text-muted text-center">No todos left</h5>
                    </>
                    : <>
                      <h1 className="fas fa-pencil-alt text-muted text-center">
                      </h1>
                  <h5 className="font-weight-light m-0 text-muted text-center">{viewArchive ? 'No todos found' : 'Add some todos...'}</h5>
                    </>}
                </div>
              }
            </>
            :
            <>
              {completed.length ?
                <div className={`card-body px-0 ${completedEditShow ? '' : 'pb-0'}`}>
                  <div className={`px-4 pt-0 ${css(styles.limitHeight)} ${completedEditShow ? '' : 'mb-4'}`}>
                    {completed.map(item => (<Checkbox key={item.id} id={item.id} topicId={id} text={item.content} clickOption={handleCompletedClick} viewArchive={viewArchive} />))}
                  </div>
                </div> : null
              }
            </>
          }
          {
            incompleteEditShow ?
              <div className={`card-body pt-0 card-option border-top`}>
                <p className="text-center py-3 mb-0">Mark
                  {Object.keys(incompleteList).filter(key => incompleteList[key]).length === 1 ? ' this' :
                    ' these ' + Object.keys(incompleteList).filter(key => incompleteList[key]).length} as...</p>
                <div className="d-flex justify-content-around align-items-center">
                  <button onClick={() => handleMarkList('completed')}
                    className={`btn btn-primary ${css(styles.radius, styles.shadow, styles.transition, styles.bounceInFast)}`}>
                    <span className="mr-1"> Done</span>
                    <span className="fas fa-check"></span>
                  </button>
                  <span>Or</span>
                  <button onClick={() => handleMarkList('delete')}
                    className={`btn btn-secondary ${css(styles.radius, styles.shadow, styles.transition, styles.bounceInFast)}`}>
                    <span className="mr-1"> Trash it</span>
                    <span className="fas fa-trash"></span>
                  </button>
                </div>
                {alert && alertAt === `edittodo${id}` ? <p className="text-center text-danger mb-0 mt-3" style={{ fontSize: '80%' }}>{alertMessage}</p> : null}
              </div> : null
          }
          {
            completedEditShow ?
              <div className={`card-body pt-0 card-option border-top`}>
                <p className="text-center py-3 mb-0">Mark
                  {Object.keys(completedList).filter(key => completedList[key]).length === 1 ? ' this' :
                    ' these ' + Object.keys(completedList).filter(key => completedList[key]).length} as...</p>
                <div className="d-flex justify-content-around align-items-center">
                  <button onClick={() => handleMarkList('incomplete')}
                    className={`btn btn-primary ${css(styles.radius, styles.shadow, styles.transition, styles.bounceInFast)}`}>
                    <span className="mr-1"> Todo</span>
                    <span className="fas fa-edit"></span>
                  </button>
                  <span>Or</span>
                  <button onClick={() => handleMarkList('delete')}
                    className={`btn btn-secondary ${css(styles.radius, styles.shadow, styles.transition, styles.bounceInFast)}`}>
                    <span className="mr-1"> Trash it</span>
                    <span className="fas fa-trash"></span>
                  </button>
                </div>
                {alert && alertAt === `edittodo${id}` ? <p className="text-center text-danger mb-0 mt-3" style={{ fontSize: '80%' }}>{alertMessage}</p> : null}
              </div> : null
          }
        </div>
      </div>
    </>
  )
}

Card.defaultProps = {
  entrance: 0.5,
  id: 0,
  completed: [],
  incomplete: [],
  title: '...',
  className: '',
}


const mapStateToProps = state => ({
  ...state.Misc
});

const mapDispatchToProps = dispatch => ({
  editTodos: (id, contents, type) => dispatch(editTodos(id, contents, type)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Card);
