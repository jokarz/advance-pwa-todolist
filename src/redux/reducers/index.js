import { combineReducers } from 'redux'
import Topic from './topicReducer.js'
import Misc from './miscReducer.js'
export default combineReducers({
  Misc, Topic
})
