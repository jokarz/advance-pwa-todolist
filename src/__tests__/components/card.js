import React from 'react'
import Card from '../../components/card/Card'
import { Provider } from 'react-redux'
import { createStore, applyMiddleware } from 'redux'
import thunkMiddleware from 'redux-thunk'
import rootReducer from '../../redux/reducers/'
import { render, fireEvent, wait, waitForElement } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

const defaultStoreData = () => ({
  Misc: {
    loading: false,
    alert: false,
    alertAt: 'topic',
    alertMessage: '...',
    success: false,
    successAt: 'nth',
    viewArchive: false
  },
  Topic: [{
    id: 'pscz1ggc9gb',
    title: 'Test',
    archive: false,
    todos: {
      completed: [],
      incomplete: []
    }
  }]
})




test('Card renders', async () => {
  const item = { ...defaultStoreData().Topic[0] }
  const store = createStore(
    defaultStoreData,
    applyMiddleware(
      thunkMiddleware
    )
  )
  const { asFragment } = render(<Provider store={store}>
    <Card entrance={1}
      id={item.id}
      key={item.id}
      completed={item.todos.completed || []}
      incomplete={item.todos.incomplete || []}
      title={item.title}
      setHideFab={false} />
  </Provider>)
  expect(asFragment()).toMatchSnapshot();
})

test('Card with 1 todo renders', async () => {
  const storeData = () => {
    const temp = JSON.parse(JSON.stringify(defaultStoreData()))
    temp.Topic[0].todos.incomplete.push({
      id: 'ba9k67d0l7',
      content: 'Testing'
    })
    return temp
  }
  const item = { ...storeData().Topic[0] }
  const store = createStore(
    storeData,
    applyMiddleware(
      thunkMiddleware
    )
  )
  const { container, getByPlaceholderText } = render(<Provider store={store}>
    <Card entrance={1}
      id={item.id}
      key={item.id}
      completed={item.todos.completed || []}
      incomplete={item.todos.incomplete || []}
      title={item.title}
      setHideFab={false} />
  </Provider>)
  //userEvent.type(getByPlaceholderText('Add another todo...'), 'testing')
  expect(container.querySelectorAll('.card .card-body.px-0.pb-0 .form-group').length).toEqual(1)
  expect(getByPlaceholderText('Add another todo...')).toBeTruthy()
})