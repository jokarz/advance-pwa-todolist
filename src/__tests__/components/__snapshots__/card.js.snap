// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders 1`] = `
<DocumentFragment>
  <div
    class="d-block d-sm-inline-block  "
    style="width: 100%;"
  >
    <div
      class="card shadow-sm border-0 radius_1fb25e-o_O-title_1bmx9ov-o_O-pointer_13xlah4-o_O-titleTransition_1eh30p4"
    >
      <div
        class="card-body p-3 text-center"
      >
        <h4
          class="mb-0 text-center select-none fontTitan_dzn0bb-o_O-textBlack_er4n8x"
        >
          Test
        </h4>
      </div>
    </div>
    <div
      class="card shadow-sm border-0 radius_1fb25e"
    >
      <div
        class="card-body pt-0 pb-2"
      />
      <div
        class="card-body pb-0 pt-4"
      >
        <div
          class="input-group "
        >
          <input
            class="form-control  "
            placeholder="Add a todo..."
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="card-body d-flex flex-column justify-content-center align-items-center"
      >
        <h1
          class=" text-muted text-center"
        >
          <span
            class="fas fa-pencil-alt"
          />
        </h1>
        <h5
          class="font-weight-light m-0 text-muted text-center"
        >
          Add some todos...
        </h5>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
